juniper_subscriptions/src/lib.rs

```diff
@@ -135,6 +135,121 @@ where
+        /// usually due to GraphQL validation errors (resolver errors are part of GQL_DATA message, and will be added as errors array)
+        #[serde(rename = "error")]
+        GqlError,
+        /// Server -> Client
+        /// Server sends this message to indicate that a GraphQL operation is done,
+        /// and no more data will arrive for the specific operation.
+        #[serde(rename = "complete")]
+        GqlComplete,
+        /// Client -> Server
+        /// Client sends this message in order to stop a running GraphQL operation execution
+        /// (for example: unsubscribe)
+        #[serde(rename = "stop")]
+        GqlStop,
+    }
+
+    /// Subscription Protocol Message Type Values over WS
+    #[allow(missing_docs)]
+    pub mod message_types {
+        pub const GQL_CONNECTION_INIT: &str = "connection_init";
+        pub const GQL_CONNECTION_ACK: &str = "connection_ack";
+        pub const GQL_CONNECTION_ERROR: &str = "connection_error";
+        pub const GQL_CONNECTION_KEEP_ALIVE: &str = "ka";
+        pub const GQL_CONNECTION_TERMINATE: &str = "connection_terminate";
+        pub const GQL_START: &str = "start";
+        pub const GQL_DATA: &str = "data";
+        pub const GQL_ERROR: &str = "error";
+        pub const GQL_COMPLETE: &str = "complete";
+        pub const GQL_STOP: &str = "stop";
+    }
+
+    /// Empty SubscriptionLifeCycleHandler over WS
+    pub enum SubscriptionState<'a, Context>
+    where
+        Context: Clone + Send + Sync,
+    {
+        /// The Subscription is at the init of the connection with the client after the
+        /// server receives the GQL_CONNECTION_INIT message.
+        OnConnection(Option<String>, &'a mut Context),
+        /// The Subscription is at the start of a operation after the GQL_START message is
+        /// is received.
+        OnOperation(&'a mut Context),
+        /// The subscription is on the end of a operation before sending the GQL_COMPLETE
+        /// message to the client.
+        OnOperationComplete(&'a Context),
+        /// The Subscription is terminating the connection with the client.
+        OnDisconnect(&'a Context),
+    }
+
+    /// Trait based on the SubscriptionServer
+    /// https://www.apollographql.com/docs/graphql-subscriptions/lifecycle-events/
+    pub trait SubscriptionStateHandler<Context, E>
+    where
+        Context: Clone + Send + Sync,
+        E: std::error::Error,
+    {
+        /// This function is called when the state of the Subscription changes
+        /// with the actual state.
+        fn handle(&self, _state: SubscriptionState<Context>) -> Result<(), E> {
+            Ok(())
+        }
+    }
+
+    /// A Empty Subscription Handler
+    #[derive(Default)]
+    pub struct EmptySubscriptionHandler;
+
+    impl<Context> SubscriptionStateHandler<Context, std::io::Error> for EmptySubscriptionHandler where
+        Context: Clone + Send + Sync
```

Comment:
I don't think you need clone here if it is `Send + Sync` ?

Reply:
When i was trying to develop the integration between actix web i have found some small questions about some things that seems to be possible with this integration, such as the same websocket connection for a client can have more than a subscription connection with the graphql server in a way that the apollo-client for example reuses the same websocket connection to handle more than a subscription. So i had this question, "each one of these subscription connections should share the same context or they should have isolated ones?".
Since i had considered that they should have isolated ones, i had defined that the Context in the subscriptions in actix-web should be Clone and the same i did for Context in the StateHandler, but thinking about it now it seems that they could share the same Context.